resources/analysisTools/bam2fastq/bam2Fastq.sh

```diff
@@ -7,96 +7,265 @@
+                ## See "bamListReadGroups.sh".
+                cat /dev/null | gzip -c - > "$tgtName"
+            fi
+        done
+    done
+}
+
+
+processPairedEndWithoutReadGroupsPicard() {
+    local baseName=$(basename "$FILENAME_BAM" .bam)
+
+    ## Extract the output directory from the first lane file path entered.
+    local outputDir="$outputAnalysisBaseDirectory/${baseName}_temp"
+    registerTmpFile "$outputDir"
+
+    ## Write just 2-3 FASTQs, depending on whether unpairedFastq is true.
+    ## We need to add the suffix here such that Picard automatically does the compression.
+    local fastq1BaseName="${baseName}_r1.${FASTQ_SUFFIX}"
+    local tmpFastq1=$(makeTempName "$outputDir" "$fastq1BaseName")
+    local fastq2BaseName="${baseName}_r2.${FASTQ_SUFFIX}"
+    local tmpFastq2=$(makeTempName "$outputDir" "$fastq2BaseName")
+    local PICARD_OPTIONS="$PICARD_OPTIONS COMPRESS_OUTPUTS_PER_RG=$compressIntermediateFastqs FASTQ=$tmpFastq1 SECOND_END_FASTQ=$tmpFastq2"
+    if [[ "${writeUnpairedFastq:-false}" == true ]]; then
+        local fastq3BaseName="${baseName}_r3.${FASTQ_SUFFIX}"
+        local tmpFastq3=$(makeTempName "$outputDir" "$fastq3BaseName")
+        local PICARD_OPTIONS="$PICARD_OPTIONS UNPAIRED_FASTQ=$tmpFastq3"
+    fi
+
+    ## Only process the non-supplementary reads (-F 0x800). BWA flags all alternative alignments as supplementary while the full-length
+    ## reads are exactly the ones not flagged supplementary.
+    local JAVA_OPTIONS="${JAVA_OPTIONS:-$JAVA_OPTS}"
+    "$SAMTOOLS_BINARY" view -u $(samtoolsExclusions) "$FILENAME_BAM" \
+        | "$PICARD_BINARY" $JAVA_OPTIONS SamToFastq $PICARD_OPTIONS INPUT=/dev/stdin
+
+    ## Now make sure that the output files of Picard are renamed correctly.
+    mv "$tmpFastq1" "${FILENAME_UNSORTED_FASTQ[0]}" || throw 10 "Could not move file '$tmpFastq1' to '${FILENAME_UNSORTED_FASTQ[0]}'"
+    mv "$tmpFastq2" "${FILENAME_UNSORTED_FASTQ[1]}" || throw 10 "Could not move file '$tmpFastq2' to '${FILENAME_UNSORTED_FASTQ[1]}'"
+    if [[ "${writeUnpairedFastq:-false}" == true ]]; then
+        mv "$tmpFastq3" "${FILENAME_UNSORTED_FASTQ[2]}" || throw 10 "Could not move file '$tmpFastq3' to '${FILENAME_UNSORTED_FASTQ[2]}'"
+    fi
+}
+
+processSingleEndWithReadGroupsPicard() {
+    throw 1 "processSingleEndWithReadGroups not implemented"
+}
+
+processSingleEndWithoutReadGroupsPicard() {
+    throw 1 "processSingleEndWithoutReadGroups not implemented"
+}
+
+
+
+main() {
+    setUp_BashSucksVersion
+
+    "$TOOL_BAM_IS_COMPLETE" "$FILENAME_BAM"
+
+    # Re-Array the filenames variable (outputs). Bash does not transfer arrays properly to subprocesses. Therefore Roddy encodes arrays as strings
+    # with enclosing parens. That is "(a b c)", with spaces as separators.
+    declare -ax FILENAME_UNSORTED_FASTQ=${FILENAME_UNSORTED_FASTQ}
+    declare -ax readGroups=${readGroups}
+
+    FASTQ_SUFFIX=$(getFastqSuffix)
+
+
+    if [[ "${pairedEnd:-true}" == true ]]; then
```

Comment:
Reduce indent?

``` bash
if [[ "${pairedEnd:-true}" == true && "${outputPerReadGroup:-true}" == true ]]; then
    if [[ "$converter" == "picard" ]]; then
        processPairedEndWithReadGroupsPicard
    elif [[ "$converter" == "biobambam" ]]; then
        processPairedEndWithReadGroupsBiobambam
    else
        throw 10 "Unknown bam-to-fastq converter: '$converter'"
    fi
elif [[ "${pairedEnd:-true}" ]]; then
    processPairedEndWithoutReadGroupsPicard
elif [[ "${outputPerReadGroup:-true}" == true ]]; then
    processSingleEndWithReadGroupsPicard
else
    processSingleEndWithoutReadGroupsPicard
fi
```

Reply:
Not sure that reducing the nesting (that's what you meant here, or not) reduces the complexity. I did choose the presented way because it just makes the complexity apparent, because it is just a decision tree. If I'd used `if`s with conditions representing mixed semantics in the same condition or in different conditions the levelling would be lost. E.g. this now has simple three semantic levels:

pairedEnd, outputPerReadGroup, converter

and outputtPerReadGroup is at the second level in each pairedEnd-branch.

However, note that the single-end branch is not tested and might not work. Until it is tested I could just throw an error in that branch or even just drop the condition. Not sure. The branching was decided on early (maybe too early, alas).